utils/probing.py

```diff
@@ -0,0 +1,211 @@
+        _, cache = model.run_with_cache(tokens[i:i+mb_size], return_type=None)
+        layer_list = []
+        for l in range(model.cfg.n_layers):
+            layer_act = cache[f'blocks.{l}.{component}'][:, pos, :]
+            layer_list.append(layer_act.unsqueeze(1).cpu().numpy())
+        layer_list: Float[np.ndarray, "batch layer dim"] = np.concatenate(layer_list, axis=1)
+        batch_list.append(layer_list)
+    batch_list: Float[np.ndarray, "batch layer dim"] = np.concatenate(batch_list, axis=0)
+    print(f"Activation shape: {batch_list.shape}")
+    # save as pkl file
+    with open(f"data/cached_activations/2_8b_mood_inference/{model.cfg.model_name}_{component}_pos_{pos}_activations.pkl", "wb") as f:
+        pickle.dump(batch_list, f)
+
+    return batch_list
+
+
+def train_probe_at_layer_pos(
+    model: HookedTransformer, act_folder: str, labels: Int[np.ndarray, "batch"], layer: int, pos: int = 0, 
+    component: str = 'hook_resid_post', with_scaler: bool = True, max_iter: int = 100,
+    test_size: float = 0.2, random_state: int = 42,
+) -> Tuple[Pipeline, float]:
+    """
+    Train a logistic regression probe on a given layer and position.
+    """
+    # Get activations
+    with open(os.path.join(act_folder, f"{model.cfg.model_name}_{component}_pos_{pos}_activations.pkl"), "rb") as f:
+        act_list: Float[Tensor, "batch d_model"] = pickle.load(f)[:, layer, :]
+
+    X_train, X_test, y_train, y_test = train_test_split(act_list, labels, test_size=test_size, random_state=random_state)
+
+    if with_scaler:
+        pipe = make_pipeline(StandardScaler(), LogisticRegression(max_iter=max_iter))
+    else:
+        pipe = make_pipeline(LogisticRegression(max_iter=max_iter))
+    pipe.fit(X_train, y_train)  # apply scaling on training data
+    score = pipe.score(X_test, y_test)  # apply scaling on testing data, without leaking training data.
+    print(f"Layer {layer} position {pos} test score: {score}")
+
+    return pipe, score
+
+
+def get_probe_direction(pipeline: Pipeline, with_scaler: bool = True) -> Tuple[float, float]:
+    """
+    Extracts the direction in the residual stream corresponding to a fitted logistic regression probe.
+    """
+    if with_scaler:
+        # Get the standard deviation and mean from the StandardScaler
+        scaler = pipeline.named_steps['standardscaler']
+        std = scaler.scale_
+        mean = scaler.mean_
+
+        # Get the coefficients and intercept from the LogisticRegression
+        logistic_regression_model = pipeline.named_steps['logisticregression']
+        scaled_coef = logistic_regression_model.coef_
+        scaled_intercept = logistic_regression_model.intercept_
+
+        # Unscale the coefficients
+        probe_coef = scaled_coef / std
+
+        # Unscale the intercept
+        probe_intercept = scaled_intercept - (scaled_coef * mean / std).sum(axis=1)
+    else:
+        logistic_regression_model = pipeline.named_steps['logisticregression']
```

Comment:
Minor: three duplicated lines from above

Reply:
Restructured this to remove repetition and inefficient ordering